src/mono/mono/utils/mono-threads-posix.c

```diff
@@ -282,6 +282,13 @@ mono_native_thread_set_name (MonoNativeThreadId tid, const char *name)
 	if (tid != mono_native_thread_id_get ())
 		return;
 
+	/*
+	 * Ignore requests to set the main thread name because
+	 * it causes the value returned by Process.ProcessName to change.
+	 */
+	if (tid == getpid())
+		return;
```

Comment:
```suggestion
#if defined(__linux__)
		if (mono_native_thread_os_id_get () == (guint64)getpid ())
			return;
#endif
```

I think this would work.

Reply:
If it doesn't work, please disable the test for Mono and open an issue for us

Reply:
I don't have a Linux machine in a good state to verify it right now, but it does no harm to push and see...

Reply:
`getpid` and `pthread_t` (aka `MonoNativeThreadId`) can't be compared directly. On Linux, `pthread_t` points to some user-space struct about the thread. While `getpid` is from pid namespace that is maintained by the kernel. For the `pthread_t` there is a corresponding `pid_t`, but there is no public API to get it (from a `pthread_t`). Instead, we can call Linux specific `gettid` when the thread is started and store that value, so we can use it later for this comparison.

I can look at changing the mono implementation. I won't get to it this week.

Reply:
@tmds That's why I suggested `mono_native_thread_os_id_get ()` which just returns the tid on Linux.